# Week 3 machine learning


## Trying to distinguish male and female voice based on data

### Data cleaning

In [1]:
import pandas as pd
import seaborn as sns
import sklearn as sk

In [2]:
df = pd.read_csv("voice.csv")

#Subset variables so we only use the ones we need
subset = df[["meanfreq","sd","median","sfm","meanfun","meandom","modindx","label"]]

#Turn male/female into a dummy variable
dummies = pd.get_dummies(subset["label"])
subset = pd.concat([subset,dummies], axis = 1)
subset.head()

meanfreq        sd    median       sfm   meanfun   meandom   modindx label  \
0  0.059781  0.064241  0.032027  0.491918  0.084279  0.007812  0.000000  male   
1  0.066009  0.067310  0.040229  0.513724  0.107937  0.009014  0.052632  male   
2  0.077316  0.083829  0.036718  0.478905  0.098706  0.007990  0.046512  male   
3  0.151228  0.072111  0.158011  0.727232  0.088965  0.201497  0.247119  male   
4  0.135120  0.079146  0.124656  0.783568  0.106398  0.712812  0.208274  male   

   female  male  
0       0     1  
1       0     1  
2       0     1  
3       0     1  
4       0     1

### Predictive model

The random forest algorithm creates several small neural networks called tree's. Each tree gets 1 vote and the decision that gets the most votes passes. 

In [3]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

# Split the data into training and test sets
X = subset[subset.columns[:7]]
y = subset["label"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

#Fit the model

clf = RandomForestClassifier(n_estimators=100)
clf = clf.fit(X_train, y_train)

### Evaluation

In [4]:
from sklearn.metrics import confusion_matrix

#Test the fit by making a confusion matrix
y_test_pred = clf.predict(X_test)
cm = confusion_matrix(y_test, y_test_pred) 
cm

array([[442,  15],
       [ 14, 480]], dtype=int64)

In [5]:
# add labels
conf_matrix = pd.DataFrame(cm, index=["Female", "Male", ], columns = ["Female_p",  "Male_p"]) 
conf_matrix

Female_p  Male_p
Female       442      15
Male          14     480

### Conclusion

Female: Precision = 442 / 456 =  97%
        Recall = 442 / 457 = 97%
        
Male: Precision = 480 / 495 = 97%
      Recall = 480/494 = 97%
      
The model is 97% accurate